gc/base/TLHAllocationSupport.cpp

```diff
@@ -419,4 +431,55 @@ MM_TLHAllocationSupport::objectAllocationNotify(MM_EnvironmentBase *env, void *h
 }
 #endif /* OMR_GC_OBJECT_ALLOCATION_NOTIFY */
 
+/**
+ * Reserve part (top) of TLH for GC if collector requires
+ */
+void
+MM_TLHAllocationSupport::reserveTLHTopForGC(MM_EnvironmentBase *env)
+{
+	uintptr_t bytesToReserveForGC = env->getExtensions()->getGlobalCollector()->reservedForGCAllocCacheSize();
+
+	Assert_MM_true(_reservedBytesForGC == 0);
+
+	if (bytesToReserveForGC > 0) {
+		_reservedBytesForGC = bytesToReserveForGC;
+		setTop((void *) ((uintptr_t) getTop() - _reservedBytesForGC));
+	}
+}
+
+/**
+ * Restore TLH Top if it was reserved for GC during refresh, and then notify the collector.
+ *
+ * TLH Top must be restored before it's flushed/cleared.
+ */
+void
+MM_TLHAllocationSupport::restoreTLHTopForGC(MM_EnvironmentBase *env)
+{
+	MM_GCExtensionsBase *extensions = env->getExtensions();
+
+	if (NULL == getBase()) {
+		Assert_MM_true(NULL == getTop());
+		Assert_MM_true(0 == _reservedBytesForGC);
+
+	} else {
+		Assert_MM_true(NULL != getTop());
+
+		if (_reservedBytesForGC > 0) {
+			Assert_MM_true(extensions->usingSATBBarrier());
+			Assert_MM_true(_reservedBytesForGC == extensions->getGlobalCollector()->reservedForGCAllocCacheSize());
+
+			setTop((void *)((uintptr_t) getTop() + _reservedBytesForGC));
+			_reservedBytesForGC = 0;
+
+			if (getUsedSize() > 0) {
+				void *lastTLHobj = getAlloc();
+
+				extensions->objectModel.initializeMinimumSizeObject(env, lastTLHobj);
+				setAlloc((void *)((uintptr_t) lastTLHobj + OMR_MINIMUM_OBJECT_SIZE));
+
+				(extensions->getGlobalCollector())->allocCacheRestored(env, getBase(), lastTLHobj);
```

Comment:
Restored in allocCacheRestored is questionable
From caller's point of veiew make sense, but from callee's point of view not so. Restored from what? Collector does not know anything about reserve/restore game.
Some more generic term like Flushed could be appropriate. Although even that name has a problem, but from caller's point of view, since it's not on common TLH flush path.
But it could be, just after we return from restoreTLHTopForGC, in which case allocCacheRestored/Flushed should have a filter inside 'if SATB active'.
Another hybrid approach is to have allocCacheRestored retrun bool (if _reservedBytesForGC > 0) and only conditionally call allocCacheFlushed. 

Think about it...

In future, if we do _pre_-marking apprach, that method name would probalby change again to something like allocCacheCreated

Reply:
Somewhat indecisive how to proceed with this, it might be best to just remove collector call here and directly call `markObjectsForRange`.

Reply:
Code gets a bit awkward with these approches:

>But it could be, just after we return from restoreTLHTopForGC, in which case allocCacheRestored/Flushed should have a filter inside 'if SATB active'.
Another hybrid approach is to have allocCacheRestored retrun bool (if _reservedBytesForGC > 0) and only conditionally call allocCacheFlushed.

Reply:
> it might be best to just remove collector call here and directly call markObjectsForRange.

The collector call doesn't serve much of a purpose anymore ever since we moved most of things out from their into TLH. Only thing now is that it's just a wrapper to  `markObjectsForRange`

Reply:
>Code gets a bit awkward with these approches

to clarify what I mean by this... `lastTLHobj` is required to call `allocCacheRestored`:

```C++

void *lastTLHobj = getAlloc();

extensions->objectModel.initializeMinimumSizeObject(env, lastTLHobj);
setAlloc((void *)((uintptr_t) lastTLHobj + OMR_MINIMUM_OBJECT_SIZE));

 extensions->getGlobalCollector()->allocCacheRestored(env, getBase(), lastTLHobj);
```

Reply:
> > Code gets a bit awkward with these approches
> 
> to clarify what I mean by this... `lastTLHobj` is required to call `allocCacheRestored`:
> 
> ```c++
> void *lastTLHobj = getAlloc();
> 
> extensions->objectModel.initializeMinimumSizeObject(env, lastTLHobj);
> setAlloc((void *)((uintptr_t) lastTLHobj + OMR_MINIMUM_OBJECT_SIZE));
> 
>  extensions->getGlobalCollector()->allocCacheRestored(env, getBase(), lastTLHobj);
> ```

lastTLHobj could be returned by restoreTLHTopForGC, either as return value or as an output parameter.
But I don't want to be pushy... give it one more though, settle with something, even if it's what currently in PR, and let's complete this.

Reply:
What I have now is:
- `restoreTLHTopForGC` returns `lastTLHobj`, which is null checked before invoking collector
- Renamed `allocCacheRestored` to `preAllocCacheWipe`... "Wipe" seems to be common between the two paths that it's called on (the flush path and refresh path). If you have another name I can change it or if we should change it to `allocCacheRestoredFlushed`?

Reply:
looks reasonable to me as it is now